src/sentry/monitors/logic/mark_failed.py

```diff
@@ -117,7 +117,9 @@ def mark_failed_threshold(failed_checkin: MonitorCheckIn, failure_issue_threshol
         # use .values() to speed up query
         previous_checkins = list(
             reversed(
-                MonitorCheckIn.objects.filter(monitor_environment=monitor_env)
+                MonitorCheckIn.objects.filter(
+                    monitor_environment=monitor_env, date_added__lte=failed_checkin.date_added
+                )
```

Comment:
This doesn't seem like the change we actually want to make here, since now the monitor could go into a failed state when it is actually already recovered no? We __always__ want to look at the most recent check-ins I think. But I think we need to **ignore in progress check-ins** here

Reply:
I think if a timeout occurs even in the past we should send an event/occurrence just like we do now.

Reply:
I'd rather just fix this now and then have a separate discussion after about design decisions. Or just wait for more feedback from customers

Reply:
This change doesn't appear to be correct though.

The problem I can imagine is, let's say we have a threshold of 4

If it looks like this

```
X X I X I I

X = failed
I = in-progress
```

When the 3rd check in fails we will **not** detect a incident, because we're looking back starting at the 3rd checkin, we need to look at the most recent check-ins not in progress right?

that said, there might still be a problem here

```
X I I I X X I I
```

If we're ignoring in progress check ins, and the 2nd to last in-progress is marked as failed, we would create an incident, but that doesn't really make sense right, those 3 in progress check ins at the start may be OK.

So I think what we want to actually do is find the most-recent NOT in-progress check-in and use that as the filter date parameter.

Reply:
Makes sense since we're indexed. I'll update the PR

Reply:
Wait I think this change doesn't address the second scenario I mentioned here.

We can't just filter out in-progress, since we shouldn't consider spaced-out failed check-ins as an incident. since what might start like this

```
X X I X X
```

could very well turn into this

```
X X O X X

X = failed
I = in-progress
O = OK
```

What we want to do is yes include in-progress check-ins in our list. But we need to start our list from the most recent terminated check in (not in progress)

Something like this

```py
# Check if the icndient is recovering based on the current state
last_terminated_check_in = (
    MonitorCheckIn.objects.filter(
        MonitorEnvironment=monitor_env, status__in=CheckInStatus.FINISHED_VALUES
    )
    .values_list("date_added")
    .order_by("-date_added")[0]
)

previous_checkins = (
    MonitorCheckIn.objects.filter(
        monitor_environment=monitor_env, date_added__lte=last_terminated_check_in
    )
    .values("id", "date_added", "status")
    .order_by("-date_added")[:recovery_threshold]
)
```

Reply:
Chatted offline with @rjo100 about this.

My comment above **is** more correct in terms of behavior, but is still missing some other cases such as 

```
failure_threshold=4
recovery_threshold=4

X X I X X O O
```

In this scenario even once the in progress check-in becomes a failure, we would fail to mark this as an incident since we're only looking back 4 check ins. We would need to have a bit more advanced logic to cover this case.

So for now, we're going to merge this PR as is, with the case of in progress in the middle of failures not being covered, and incorrectly creating an incident here, BUT correctly creating an incident when there are trailing in-progress check-ins.

In the future we will cover more cases in a more comprehensive document.